leakcanary-deobfuscation-gradle-plugin/src/main/java/com/squareup/leakcanary/deobfuscation/LeakcanaryLeakDeobfuscationPlugin.kt

```diff
@@ -0,0 +1,128 @@
+package com.squareup.leakcanary.deobfuscation
+
+import com.android.build.gradle.AppExtension
+import com.android.build.gradle.BaseExtension
+import com.android.build.gradle.LibraryExtension
+import com.android.build.gradle.api.ApplicationVariant
+import com.android.build.gradle.api.BaseVariant
+import com.android.build.gradle.api.LibraryVariant
+import org.gradle.api.DefaultTask
+import org.gradle.api.DomainObjectSet
+import org.gradle.api.GradleException
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.Task
+import org.gradle.api.UnknownTaskException
+import org.gradle.api.logging.LogLevel.DEBUG
+import org.gradle.api.tasks.TaskProvider
+import java.io.File
+
+class LeakcanaryLeakDeobfuscationPlugin : Plugin<Project> {
+
+  override fun apply(project: Project) {
+    val variants = findAndroidVariants(project)
+    val leakCanaryExtension = createLeakCanaryExtension(project)
+    variants.all { variant ->
+      if (leakCanaryExtension.filterObfuscatedVariants(variant)) {
+        setupTasks(project, variant)
+      }
+    }
+  }
+
+  private fun findAndroidVariants(project: Project): DomainObjectSet<BaseVariant> {
+    return when (val extension = project.extensions.getByType(BaseExtension::class.java)) {
+      is AppExtension -> extension.applicationVariants as DomainObjectSet<BaseVariant>
+      is LibraryExtension -> extension.libraryVariants as DomainObjectSet<BaseVariant>
+      else -> throwNoAndroidPluginException()
+    }
+  }
+
+  private fun createLeakCanaryExtension(project: Project): LeakCanaryDeobfuscationExtension {
+    return project.extensions.create("leakCanary", LeakCanaryDeobfuscationExtension::class.java)
+  }
+
+  private fun setupTasks(
+    project: Project,
+    variant: BaseVariant
+  ) {
+    val copyObfuscationMappingFileTaskProvider = project.tasks.register(
+        "leakCanaryCopyObfuscationMappingFor${variant.name.capitalize()}",
+        CopyObfuscationMappingFileTask::class.java
+    ) {
+      it.mappingFile = getMappingFile(project, variant)
+      it.mergeAssetsDirectory = variant.mergeAssetsProvider.get().outputDir.get().asFile
+
+      val mappingGeneratingTaskProvider =
+        findTaskProviderOrNull(
+            project,
+            "transformClassesAndResourcesWithR8For${variant.name.capitalize()}"
+        ) ?: findTaskProviderOrNull(
+            project,
+            "transformClassesAndResourcesWithProguardFor${variant.name.capitalize()}"
+        ) ?: throwMissingMinifiedVariantException(project)
+
+      it.dependsOn(mappingGeneratingTaskProvider)
+    }
+
+    getPackageTaskProvider(variant).configure {
+      it.dependsOn(copyObfuscationMappingFileTaskProvider)
+    }
+  }
+
+  private fun getMappingFile(
+    project: Project,
+    variant: BaseVariant
+  ): File {
+    return variant.mappingFile ?: throwMissingMappingFileForVariantException(project, variant.name)
```

Comment:
Here you check the mappingFile existence for 1 variant but the error message refers to all variants. It is not clear.
Moreover, we already perform a similar check in the task itself, which is better (because a task could be configured outside the plugin).  So maybe here, we can just pass the mappingFile of the variant to the task and in the task make the property nullable, and add a test to see if it is not null in the checks that the tasks perform. Maybe change the error message of the check there to say "The plugin was configured to be applied to the variant <name> but this variant doesn't define an obfuscation mapping file: make sure the variant. isMinified is true".